### PR TITLE
Number -> Real

### PR DESCRIPTION
**Pull request by YingboMa**
Much code uses `Real` as type constraints so maybe we should use `Real` instead of `Number` to be able to trace through more code.

CC: @jlperla

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
What about going crazy and using AbstractFloat?

**Comment by isaacsas** [comment]
Would this break things like
```julia
@parameters n
b = binomial(n,2)
```

**Comment by ChrisRackauckas** [comment]
Nope, it'll only superset the number of functions we can trace into (since being a stricter subtype makes you still a subtype of the supertype). So in a way, AbstractFloat might just be better than Real and better than Number if the only goal is to just trace whatever we can.

**Comment by jlperla** [comment]
Yeah, AbstractFloat might be better.

**Comment by YingboMa** [comment]
> What about going crazy and using AbstractFloat?

I don't think there is much code uses `::AbstractFloat` signature that's not an elementary/special function, so using `AbstractFloat` won't help us much.

**Comment by ChrisRackauckas** [comment]
But if it allows strictly more code to trace, is it harmful? I think it might as well be done?

File: src/systems/jumps/jumpsystem.jl [hunk 1]
```diff
@@ -154,6 +176,40 @@ function DiffEqBase.DiscreteProblem(sys::JumpSystem, u0map, tspan::Tuple,
     DiscreteProblem(df, u0, tspan, p; kwargs...)
 end
 
+"""
+```julia
+function DiffEqBase.DiscreteProblemExpr(sys::JumpSystem, u0map, tspan,
+                                    parammap=DiffEqBase.NullParameters; kwargs...)
+```
+
+Generates a black DiscreteProblem for a JumpSystem to utilize as its
```
Comment (isaacsas) [inline src/systems/jumps/jumpsystem.jl:185]: We should fix that typo "black" -> "blank".

**Pull request closed without merging**